File: docs/snippets/all/tutorials/custom_data.cpp
/// Demonstrates how to implement custom archetypes and components, and extend existing ones.

#include <dalaran.hpp>
#include <dalaran/demo_utils.hpp>

/// A custom component type.
struct Confidence {
    float value;
};

template <>
struct dalaran::Loggable<Confidence> {
    static constexpr std::string_view ComponentType = "user.Confidence";

    static const std::shared_ptr<arrow::DataType>& arrow_datatype() {
        return dalaran::Loggable<dalaran::Float32>::arrow_datatype();
    }

    // TODO(#4257) should take a dalaran::Collection instead of pointer and size.
    static dalaran::Result<std::shared_ptr<arrow::Array>> to_arrow(
        const Confidence* instances, size_t num_instances
    ) {
        return dalaran::Loggable<dalaran::Float32>::to_arrow(
            reinterpret_cast<const dalaran::Float32*>(instances),
            num_instances
        );
    }
};

/// A custom archetype that extends Dalaran's builtin `dalaran::Points3D` archetype with a custom component.
struct CustomPoints3D {
    dalaran::Points3D points;
    // Using a dalaran::Collection is not strictly necessary, you could also use an std::vector for example,
    // but useful for avoiding allocations since `dalaran::Collection` can borrow data from other containers.
    std::optional<dalaran::Collection<Confidence>> confidences;
};

template <>
struct dalaran::AsComponents<CustomPoints3D> {
    static Result<dalaran::Collection<ComponentBatch>> as_batches(
        const CustomPoints3D& archetype
    ) {
        auto batches =
            AsComponents<dalaran::Points3D>::as_batches(archetype.points)
                .value_or_throw()
                .to_vector();

        // Add custom confidence components if present.
        if (archetype.confidences) {
            auto descriptor =
                dalaran::ComponentDescriptor("user.CustomPoints3D:confidences")
                    .or_with_archetype("user.CustomPoints3D")
                    .or_with_component_type(
                        dalaran::Loggable<Confidence>::ComponentType
                    );
            batches.push_back(ComponentBatch::from_loggable(
                                  *archetype.confidences,
                                  descriptor
            )
                                  .value_or_throw());
        }

        return dalaran::take_ownership(std::move(batches));
    }
};

// ---

int main(int argc, char* argv[]) {
    const auto rec = dalaran::RecordingStream("dalaran_example_custom_data");
    rec.spawn().exit_on_failure();

    auto grid = dalaran::demo::grid3d<dalaran::Position3D, float>(-5.0f, 5.0f, 3);

    rec.log(
        "left/my_confident_point_cloud",
        CustomPoints3D{
            dalaran::Points3D(grid),
            Confidence{42.0f},
        }
    );

    std::vector<Confidence> confidences;
    for (auto i = 0; i < 27; ++i) {
        confidences.emplace_back(Confidence{static_cast<float>(i)});
    }

    rec.log(
        "right/my_polarized_point_cloud",
        CustomPoints3D{
            dalaran::Points3D(grid),
            confidences,
        }
    );
}

File: docs/snippets/all/tutorials/custom_data.rs
//! Demonstrates how to implement custom archetypes and components, and extend existing ones.

use dalaran::{
    ComponentBatch as _, ComponentDescriptor, SerializedComponentBatch,
    demo_util::grid,
    external::{arrow, glam, dl_sdk_types},
};

// ---

/// A custom [component bundle] that extends Dalaran's builtin [`dalaran::Points3D`] archetype with extra
/// [`dalaran::Component`]s.
///
/// [component bundle]: [`AsComponents`]
struct CustomPoints3D {
    points3d: dalaran::Points3D,
    confidences: Option<Vec<Confidence>>,
}

impl dalaran::AsComponents for CustomPoints3D {
    fn as_serialized_batches(&self) -> Vec<SerializedComponentBatch> {
        std::iter::chain(
            self.points3d.as_serialized_batches(),
            std::iter::once(self.confidences.as_ref().and_then(|batch| {
                batch.serialized(ComponentDescriptor {
                    archetype: Some("user.CustomPoints3D".into()),
                    component: "user.CustomPoints3D:confidences".into(),
                    component_type: Some(
                        <Confidence as dalaran::Component>::name(),
                    ),
                })
            }))
            .flatten(),
        )
        .collect()
    }
}

// ---

/// A custom [`dalaran::Component`] that is backed by a builtin [`dalaran::Float32`] scalar.
#[derive(Debug, Clone, Copy, dalaran::SizeBytes)]
struct Confidence(dalaran::Float32);

impl From<f32> for Confidence {
    fn from(v: f32) -> Self {
        Self(dalaran::Float32(v))
    }
}

impl dalaran::Loggable for Confidence {
    #[inline]
    fn arrow_datatype() -> arrow::datatypes::DataType {
        dalaran::Float32::arrow_datatype()
    }

    #[inline]
    fn to_arrow_opt<'a>(
        data: impl IntoIterator<
            Item = Option<impl Into<std::borrow::Cow<'a, Self>>>,
        >,
    ) -> dl_sdk_types::SerializationResult<arrow::array::ArrayRef>
    where
        Self: 'a,
    {
        dalaran::Float32::to_arrow_opt(
            data.into_iter().map(|opt| opt.map(Into::into).map(|c| c.0)),
        )
    }
}

impl dalaran::Component for Confidence {
    #[inline]
    fn name() -> dalaran::ComponentType {
        "user.Confidence".into()
    }
}

// ---

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = dalaran::RecordingStreamBuilder::new("dalaran_example_custom_data")
        .spawn()?;

    rec.log(
        "left/my_confident_point_cloud",
        &CustomPoints3D {
            points3d: dalaran::Points3D::new(grid(
                glam::Vec3::splat(-5.0),
                glam::Vec3::splat(5.0),
                3,
            )),
            confidences: Some(vec![42f32.into()]),
        },
    )?;

    rec.log(
        "right/my_polarized_point_cloud",
        &CustomPoints3D {
            points3d: dalaran::Points3D::new(grid(
                glam::Vec3::splat(-5.0),
                glam::Vec3::splat(5.0),
                3,
            )),
            confidences: Some(
                (0..27).map(|i| i as f32).map(Into::into).collect(),
            ),
        },
    )?;

    Ok(())
}
